src/gil.rs

```diff
@@ -275,14 +269,9 @@ pub unsafe fn register_pointer(obj: NonNull<ffi::PyObject>) {
     }
 }
 
-pub unsafe fn register_owned(_py: Python, obj: NonNull<ffi::PyObject>) -> &PyAny {
+pub unsafe fn register_owned(_py: Python, obj: NonNull<ffi::PyObject>) {
     let pool = POOL.get_or_init();
-    &*(pool.owned.push_back(obj) as *const _ as *const PyAny)
-}
-
-pub unsafe fn register_borrowed(_py: Python, obj: NonNull<ffi::PyObject>) -> &PyAny {
-    let pool = POOL.get_or_init();
-    &*(pool.borrowed.push_back(obj) as *const _ as *const PyAny)
+    pool.owned.push(obj);
```

Comment:
Idea: could we make the object storage for owned pointers a `thread_local!`, rather than using the `'static` GILPool? Might be even faster and avoid potential threading issues (might even resolve #756).

Reply:
If we're going to support Python on embedded platforms (not sure), we can't use `thread_local`, however, unless we currently support embedded platforms, I have no issues requiring `thread_local` support.

Some other thing we should think about, how does it interact with futures trying to move between threads (just make them non-`Send`?), also, does it work with scoped threads?

Reply:
Also, does this force types like `Py<PyAny>` to be non-`Send`able?

Reply:
> If we're going to support Python on embedded platforms (not sure), we can't use thread_local, however, unless we currently support embedded platforms, I have no issues requiring thread_local support.

Oh really? Have you some docs on that? I never saw this issue with thread_local before.

Regarding all the `Send` interactions - I think even if `PyAny` is not `Send`, we can explicitly `impl Send for Py<T>`, right?

Reply:
> > If we're going to support Python on embedded platforms (not sure), we can't use thread_local, however, unless we currently support embedded platforms, I have no issues requiring thread_local support.
> 
> Oh really? Have you some docs on that? I never saw this issue with thread_local before.

I didn't find docs that explicitly say that, however https://github.com/rust-lang/rust/issues/62918 alludes to that:

> Futures currently have a dependency on TLS for storing a pointer to` Context<'_>`.
> https://github.com/rust-lang/rust/blob/master/src/libstd/future.rs#L116
> 
> This makes usage in `no_std` environments quite difficult.

Reply:
> Regarding all the `Send` interactions - I think even if `PyAny` is not `Send`, we can explicitly `impl Send for Py<T>`, right?

Yeah. I think that may be what we already do, though didn't check.

Reply:
regarding embedded and TLS, https://github.com/rustasync/team/issues/42 (referenced from previously mentioned issue) is much more clear.

Reply:
Oh I see; I think what you're saying is that `thread_local!` isn't available in `no_std` applications? It would be interesting to think about pyO3 support for `no_std`, but I think it's a long way off. Also not sure how many users might want this?

Reply:
In that case, I think we should just switch to `thread_local` for now, and, whenever we add `no_std` support, we could `cfg` something else in.

Reply:
>  could we make the object storage for owned pointers a thread_local!, rather than using the 'static GILPool?

Possibly we can, but I failed to implement it due to some odd errors... :thinking:

Reply:
@kngwyu I have just opened https://github.com/kngwyu/pyo3/pull/2 which implements `thread_local`